File: LeetCode/Medium/1353-Maximum_Number_Of_Events_That_Can_Be_Attended/1353-MaximumNumberOfEventsThatCanBeAttended.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <functional>
#include <algorithm>

class Solution
{
public:
    int maxEvents(vector<vector<int>> &events)
    {
        int n = events.size();
        int count = 0;
        std::sort(events.begin(), events.end(), [](const std::vector<int> &a, const std::vector<int> &b)
                  { return a[0] < b[0]; });

        int day = 0;
        int j = 0;
        std::priority_queue<int, std::vector<int>, std::greater<int>> available;

        while (j < n || !available.empty())
        {

            if (available.empty() && j < n)
                day = max(day, events[j][0]);

            while (j < n && events[j][0] == day)
            {
                available.push(events[j][1]);
                j++;
            }

            while (!available.empty() && available.top() < day)
                available.pop();

            if (!available.empty())
            {
                available.pop();
                count++;
                day++;
            }
        }

        return count;
    }
};

File: LeetCode/Medium/1353-Maximum_Number_Of_Events_That_Can_Be_Attended/1353-MaximumNumberOfEventsThatCanBeAttended.py
import heapq


class Solution:
    def maxEvents(self, events: list[list[int]]) -> int:
        n = len(events)
        count = 0
        events.sort(key=lambda x: x[0])

        day = 0
        j = 0
        available = []

        while j < n or available:

            if not available and j < n:
                day = max(day, events[j][0])

            # Add potential events to the available
            while j < n and events[j][0] == day:
                heapq.heappush(available, events[j][1])
                j += 1

            # Remove invalid events
            while available and available[0] < day:
                heapq.heappop(available)

            # pick the best event from avaliable
            if available:
                heapq.heappop(available)
                count += 1
                day += 1

        return count


a = Solution()
print(a.maxEvents(events=[[1, 2], [2, 3], [3, 4]]))
print(a.maxEvents(events=[[1, 2], [2, 3], [3, 4], [1, 2]]))
print(a.maxEvents(events=[[1, 2], [2, 2], [3, 3], [3, 4], [3, 4]]))
print(a.maxEvents(events=[[1, 2], [1, 2], [3, 3], [1, 5], [1, 5]]))
